src/REP15.sol

```diff
@@ -0,0 +1,466 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import { IERC165, ERC165 } from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import { IERC721, ERC721 } from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import { IREP15 } from "./interfaces/IREP15.sol";
+import { IREP15Errors } from "./interfaces/IREP15Errors.sol";
+import { IREP15ContextCallback } from "./interfaces/IREP15ContextCallback.sol";
+import { REP15Utils } from "./REP15Utils.sol";
+
+abstract contract REP15 is ERC721, IREP15, IREP15Errors {
+  using REP15Utils for REP15Utils.Delegation;
+  using REP15Utils for REP15Utils.Context;
+  using REP15Utils for REP15Utils.TokenContext;
+
+  uint64 internal immutable _maxDetachingDuration;
+
+  mapping(uint256 tokenId => REP15Utils.Delegation) private _delegations;
+
+  mapping(bytes32 ctxHash => REP15Utils.Context) private _contexts;
+
+  mapping(uint256 tokenId => mapping(bytes32 ctxHash => REP15Utils.TokenContext)) private _tokenContext;
+
+  mapping(uint256 tokenId => bytes32[] ctxHashes) internal _attachedContexts;
+
+  mapping(uint256 tokenId => mapping(bytes32 ctxHash => uint256 index)) internal _attachedContextsIndex;
+
+  constructor(uint64 maxDetachingDurationSeconds) {
+    _maxDetachingDuration = maxDetachingDurationSeconds;
+  }
+
+  /**
+   * @inheritdoc IERC165
+   */
+  function supportsInterface(bytes4 interfaceId) public view virtual override(ERC721, IERC165) returns (bool) {
+    return interfaceId == type(IREP15).interfaceId || super.supportsInterface(interfaceId);
+  }
+
+  /**
+   * @inheritdoc IREP15
+   */
+  function startDelegateOwnership(uint256 tokenId, address delegatee, uint64 until) external virtual {
+    address owner = _requireOwned(tokenId);
+
+    if (delegatee == owner || delegatee == address(0)) revert REP15InvalidDelegatee(delegatee);
+
+    if (until <= block.timestamp) revert REP15InvalidDelegationExpiration(until);
+
+    REP15Utils.Delegation storage delegation = _delegations[tokenId];
```

Comment:
should use storage naming pattern

Reply:
> Prefix $ is allowed but not encouraged for storage pointers.

> All storage pointers to custom slot MUST BE $ or prefix with $_

Reply:
it will cause confusing with local parameters.
Should either use `sDelegation` or `$delegation` for clearer storage declaration.

Reply:
okay